File: examples/go/player/agg/handlers/reserve.go
// DOC: This file is referenced in docs/docs/examples/aggregates.mdx
//      Update documentation when making changes to handler patterns.
package handlers

import (
	"encoding/hex"
	"time"

	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

func guardReserveFunds(state PlayerState) error {
	if !state.Exists() {
		return angzarr.NewCommandRejectedError("Player does not exist")
	}
	return nil
}

func validateReserveFunds(cmd *examples.ReserveFunds, state PlayerState) (int64, error) {
	amount := int64(0)
	if cmd.Amount != nil {
		amount = cmd.Amount.Amount
	}
	if amount <= 0 {
		return 0, angzarr.NewCommandRejectedError("amount must be positive")
	}
	if amount > state.AvailableBalance() {
		return 0, angzarr.NewCommandRejectedError("Insufficient funds")
	}

	tableKey := hex.EncodeToString(cmd.TableRoot)
	if _, exists := state.TableReservations[tableKey]; exists {
		return 0, angzarr.NewCommandRejectedError("Funds already reserved for this table")
	}

	return amount, nil
}

func computeFundsReserved(cmd *examples.ReserveFunds, state PlayerState, amount int64) *examples.FundsReserved {
	newReserved := state.ReservedFunds + amount
	newAvailable := state.Bankroll - newReserved
	return &examples.FundsReserved{
		Amount:              cmd.Amount,
		TableRoot:           cmd.TableRoot,
		NewAvailableBalance: &examples.Currency{Amount: newAvailable, CurrencyCode: "CHIPS"},
		NewReservedBalance:  &examples.Currency{Amount: newReserved, CurrencyCode: "CHIPS"},
		ReservedAt:          timestamppb.New(time.Now()),
	}
}

// HandleReserveFunds handles the ReserveFunds command (reserve funds for table buy-in).
func HandleReserveFunds(
	commandBook *pb.CommandBook,
	commandAny *anypb.Any,
	state PlayerState,
	seq uint32,
) (*pb.EventBook, error) {
	var cmd examples.ReserveFunds
	if err := proto.Unmarshal(commandAny.Value, &cmd); err != nil {
		return nil, err
	}

	if err := guardReserveFunds(state); err != nil {
		return nil, err
	}
	amount, err := validateReserveFunds(&cmd, state)
	if err != nil {
		return nil, err
	}

	event := computeFundsReserved(&cmd, state, amount)

	eventAny, err := anypb.New(event)
	if err != nil {
		return nil, err
	}

	return angzarr.NewEventBook(commandBook.Cover, seq, eventAny), nil
}

File: examples/rust/player/agg/src/handlers/reserve.rs
//! ReserveFunds command handler.
//!
//! DOC: This file is referenced in docs/docs/examples/aggregates.mdx
//!      Update documentation when making changes to handler patterns.

use angzarr_client::proto::examples::{Currency, FundsReserved, ReserveFunds};
use angzarr_client::proto::{CommandBook, EventBook};
use angzarr_client::{new_event_book, pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::state::PlayerState;

// docs:start:reserve_funds_imp
fn guard(state: &PlayerState) -> CommandResult<()> {
    if !state.exists() {
        return Err(CommandRejectedError::new("Player does not exist"));
    }
    Ok(())
}

fn validate(cmd: &ReserveFunds, state: &PlayerState) -> CommandResult<i64> {
    let amount = cmd.amount.as_ref().map(|c| c.amount).unwrap_or(0);
    if amount <= 0 {
        return Err(CommandRejectedError::new("amount must be positive"));
    }
    if amount > state.available_balance() {
        return Err(CommandRejectedError::new("Insufficient funds"));
    }

    let table_key = hex::encode(&cmd.table_root);
    if state.table_reservations.contains_key(&table_key) {
        return Err(CommandRejectedError::new("Funds already reserved for this table"));
    }

    Ok(amount)
}

fn compute(cmd: &ReserveFunds, state: &PlayerState, amount: i64) -> FundsReserved {
    let new_reserved = state.reserved_funds + amount;
    let new_available = state.bankroll - new_reserved;

    FundsReserved {
        amount: cmd.amount.clone(),
        table_root: cmd.table_root.clone(),
        new_available_balance: Some(Currency {
            amount: new_available,
            currency_code: "CHIPS".to_string(),
        }),
        new_reserved_balance: Some(Currency {
            amount: new_reserved,
            currency_code: "CHIPS".to_string(),
        }),
        reserved_at: Some(angzarr_client::now()),
    }
}

pub fn handle_reserve_funds(
    command_book: &CommandBook,
    command_any: &Any,
    state: &PlayerState,
    seq: u32,
) -> CommandResult<EventBook> {
    let cmd: ReserveFunds = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    guard(state)?;
    let amount = validate(&cmd, state)?;

    let event = compute(&cmd, state, amount);
    let event_any = pack_event(&event, "examples.FundsReserved");

    Ok(new_event_book(command_book, seq, event_any))
}
// docs:end:reserve_funds_imp
